### DISCUSSION TITLE
zrep 2.0 -- the bashening!

### DISCUSSION THREAD

**Discussion by ppbrown** [bolthole/zrep#174] [Announcements]
After a very long time... I have made the decision to change zrep to use bash!
This is for portability reasons. Getting a fully functional "proper" ksh is just too difficult these days, compared to me just converting the thing.

Currently it is on the new "zrep2.0" branch.
It  appears to work for straight sync; however, I'm sure there are bugs lurking. 
The one I'm currently stuck one:  "zrep  refresh" doesnt work, because somehow getting a lock via localhost isnt working.
The standard "kill -0 $PID" to check if $PID still is running, doesnt seem to work in the script.. yet seems to work fine when I run it by hand??  Any suggestions on that one would be welcome

But I encourage everyone to try it out. with caution :)

**Comment by dca58** [comment]
There are a few things which do not work in bash:

- date computations (printf "%(%Y-%m-%d)T\n” yesterday) and the like cannot be done within bash, you are bound to use an external program like date of which there are several different versions, and hence is not portable
- pipes are handled differently, e.g. print "1 2" | read a b; print $a $b does not print 1 2 in bash but does in ksh
- arithmetic operations like $(( sin(1.57) )) are not supported in bash 
- and a few other quirks I can think of right now

Kind regards

**Reply by ppbrown** [reply]
Thanks for the feedback!
A few things you might be interested in: 
* allegedly, bash has supportedf the (..)T syntax since bash 4.2  However, when I actually tested it on a bash4.4... it didnt work. sigh. So using 'date' is in the code now.
* Thanks a bunch for the other notes.
I recall that many years ago, someone made a fork of zrep to run in bash (for bsd or something)
But.... there are 49 known forks now, ;and I cant find it :(  was it you who did that? Do you know where that code is?

**Reply by ppbrown** [reply]
Actally. bash DOES apparently have $(( math)) now. (just maybe not trig functions)

that would explain why sync works already  :)

https://bash.cyberciti.biz/guide/Perform_arithmetic_operations

**Comment by dca58** [comment]
It does have integer arithmetic, but no float or trigonometric functions, I think.
And then, which version of bash ? Bash is improving but it is a moving target.
Personally I stay with ksh for scripting, all those goodies are in there for so long...

**Reply by ppbrown** [reply]
yeah, i'd like to. but there are multiple targets now, where ksh is a large impediment. One was an embedded system, which ships with bash but not ksh.  Adding ksh was a big deal for the zrep user.
The latest is alpine linux. which does not seem to have any viable straightforward way of getting proper ksh.
(curse those "pdksh" people for doing a half-.. job !)

**Reply by ppbrown** [reply]
PS: bash variable handling and printing is wierd.
my latest challenge is that in one of the function back-and-forths now, I've getting spurious newline in the output, where ksh does not have that issue.
Sigh...

**Comment by ppbrown** [comment]
Huh. I thought I'd posted an update in this thread, but mayhbe I did elsewhere.

I think I have the normal push-sync working. So beta testers for that would be awesome.

Its breaking for the pull mode, aka refresh.
assistance on debugging that, wouild be greatly appreciated, since I'm short on time these days.
I think something about the invokation syntax is corrupting the  (ssh zfs send) stream, but have no idea what.
Perhaps it is related to the "spurious newline" problem I observed above. dunno.

**Comment by ppbrown** [comment]
Good update!
Thanks to [ngoralski](https://github.com/bolthole/zrep/issues?q=is%3Apr+author%3Angoralski) I was able to make a little progress.
and then, I finally noticed something holding up EvERYTHING...
somewhere in zrep2.0, there is a bug that outputs a pointless newline.
even doing
"zrep -h"
has a newline in the FRONT of everything.

So if that can be found, I think we will be pretty much there.

**Comment by ppbrown** [comment]
HI folks....

I THINK... THis may actually be DONE!


To try out pure-bash version of zrep, you can grab from

https://github.com/bolthole/zrep/blob/zrep2/zrep

Please let me know how it works for you, and any bugs you may find... or if you dont find any at all :)

**Comment by ppbrown** [comment]
I think that it has had enough testing, that I made the official release!
locking this thread. Please see official announcement thread if you want to add any v2.0 comments after this.

Announcement thread: -> https://github.com/bolthole/zrep/discussions/192
